blog/_posts/2018-04-12-in-a-nutshell.md

```diff
@@ -0,0 +1,95 @@
+
+The information present in TASTY trees can be used for many purposes.
+
+ - The compiler uses it to support separate compilation.
+ - A language server for an IDE uses it to support hyperlinking, command completion, or documentation.
+ - A build tool can use it to cross-build on different platforms and migrate code from one
+   version to another.
+ - Optimizers and analyzers can use it for deep code analysis and advanced code generation
+
+Of these usages, the first two work today. The other two are worthwhile possibilities to pursue in the future.
+
+OK, but what is Tasty _exactly_? An up-to-date version of the Tasty file format is described in file
+[TastyFormat.scala](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala)
+of the `dotc` compiler for Scala 3.
+
+## What Does it Have to Do with Macros?
+
+It turns out that Tasty also makes an excellent foundation for a new generation of reflection-based macros.
+
+The problem with the current `scala.reflect` macros is that they that they are completely dependent on the current Scala compiler (internally named `nsc`). In fact, `scala.reflect` macros are nothing but a thin veneer on top of `nsc` internals. This makes them very powerful but also fragile and hard to use. Because if this, they have had "experimental" status for their whole lifetime. Since Scala 3 uses a different compiler (`dotc`), the old reflect macros cannot be ported to it, so we need something different, and hopefully better.
+
+Another way to look at `scala.reflect` macros was that they were lacking _foundations_. Scala 3 has already a meta programming facility, with rock-solid foundations. [Principled Meta Programming] (http://dotty.epfl.ch/docs/reference/principled-meta-programming.html) is a way to support staging by adding just two operators to the language: Quote (`'`) to represent code expressions, and splice (`~`) to insert one piece of code in another. The inspiration for our approach [comes from temporal logic](https://ieeexplore.ieee.org/abstract/document/561317/).  A somewhat similar system
+is used for staging in [Meta OCaml](http://okmij.org/ftp/ML/MetaOCaml.html).
+We obtain a very high level _macro system_ by combining the two temporal operators `'` and `~` with Scala 3's `inline` feature. In a nutshell:
```

Comment:
there's an extra space between `[Principled Meta Programming]` and `(http://dotty.epfl.ch/docs/reference/principled-meta-programming.html)`, preventing markdown from recognizing it properly.